cmd/dagger/engine.go

```diff
@@ -60,10 +64,82 @@ func withEngine(
 		}
 		defer sess.Close()
 
+		// Automatically serve the module in the context directory if available.
+		if params.ServeModule {
+			cwd, err := os.Getwd()
+			if err != nil {
+				return fmt.Errorf("failed to get current working directory: %w", err)
+			}
+
+			// Check if the current context directory contains a dagger.json file.
+			configFound, err := doDaggerConfigExist(cwd)
+			if err != nil {
+				return fmt.Errorf("failed to check if current working directory contains a dagger.json file: %w", err)
+			}
+
+			if !configFound {
+				// If not, directly execute the function.
+				return fn(ctx, sess)
+			}
+
+			mod, exist, err := initializeClientGeneratorModule(ctx, sess.Dagger(), ".")
+			if err != nil {
+				return fmt.Errorf("failed to initialize current module: %w", err)
+			}
+
+			if exist {
+				for _, dep := range mod.Dependencies {
+					err := dep.AsModule().Serve(ctx)
+					if err != nil {
+						return fmt.Errorf("failed to serve dependency %w", err)
+					}
+				}
+			}
+		}
+
 		return fn(ctx, sess)
 	})
 }
 
+// doDaggerConfigExist searches for dagger.json recursively in parent directories until
+// it finds a dagger.json file, reaches the context directory or reaches the root directory.
+// This is very similar to the findUp logic of the engine but it only looks for the
+// file existence prior to loading the module to not create unnecessary operation on the
```

Comment:
I wouldn't worry about unnecessary operations, it's fast enough to not be a bottleneck. We don't want to duplicate logic like this because it makes this whole system much harder to understand and maintain.

We should use the engine API for this, if there's anything about it that needs to be adjusted let me know and we can figure it out.

Reply:
Yeah there's an explanation why, in case the configuration does not exist, when used with Dagger session, we do not want to load the source root dir: https://github.com/dagger/dagger/blob/c1468fcd4e0a8d011696d29ddfe7b6437a120e24/core/schema/modulesource.go#L341

So I could add an extra argument in `moduleSource` like `skipLoadRootDirIfConfigMissing` but I thought thought it was a bit too big and specialized for that use case so I added this middleware func to avoid that.
Let me know what your thought

Reply:
That behavior was fixed here: https://github.com/dagger/dagger/pull/9659 so shouldn't be an issue anymore

Reply:
True! Fixed in [6236eae](https://github.com/dagger/dagger/pull/9531/commits/6236eae1f741f803eb29548448d9444773fc7682)